### PR TITLE
More exposition on motivation and use cases.

### PR DESCRIPTION
**Pull request by vttale**
Need to fill in some refs, but wanted to get in the edits I made during my flight

### CODE REVIEW

File: draft-dnsop-deleg-transport.md [hunk 1]
```diff
@@ -208,31 +217,33 @@ alpn=doq tlsa=disabled mandatory=alpn
 ~~~
 {: title="tlsa SvcParam Examples for DELEG"}
 
-## Deployment Considerations
+## Resolution procedure
 
-The DELEG and SVCB records intends to replace the NS record while also adding additional functionality in order to support additional transports for the DNS. Below are discussions of considerations for deployment.
+As specified in {{?I-D.draft-dnsop-deleg}}, the ServiceMode DELEG records provide a weak signal from the operator about which transport mechanisms they would most prefer clients to use, but it is not mandatory that resolvers try the delegated servers in priority order.  It might not even be possible for a resolver to do, such as if the highest priority record specifies an alpn that the resolver does not implement, but a lower priority record has an alpn that it can use.
```
Comment (fl1ger) [inline draft-dnsop-deleg-transport.md:226]: I am still not sure if we actually want priority.

File: draft-dnsop-deleg-transport.md [hunk 2]
```diff
@@ -100,76 +100,82 @@ contributor:
...
-In this example, the authoritative nameserver is delegating using DNS-over-TLS.
-Like in SVCB, DELEG also offer the ability to use the AliasMode delegation. The example below shows an example where example.com is being delegated with a DELEG AliasMode record which can then be further resolved using standard SVCB to locate the actual parameters.
+The alpn attribute specifies DNS-over-TLS, with the tlsa attribute
+providing the necessary cryptographic material to establish the
+connection.  The ipv4hint and ipv6hint serves the traditional role of
+glue, since the ns2 server lives in the example.com zone.
 
-    example.com.  86400  IN DELEG 0   config2.example.net.
-    example.com.  86400  IN NS     ns1.example.net.
+The Additional section would continue to serve glue records for the
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:146]: This is intriguing but I don't quite believe it.  DELEG is in the parent zone, and is only retrieved by an explicit query (QTYPE=DELEG).  Sure, the parent should serve glue records alongside an NS record Answer, but why should it serve glue records alongside a DELEG Answer?

There's another wrinkle here though: SVCB recommends putting A+AAAA records for TargetName in the Additional section if they are in the same zone, to avoid a followup query.  If we apply that to DELEG, it could mean adding A+AAAA records for "ns2.example.com" (in the above example).  However, this would be redundant with the IP hints, and it would also be logically different: those A+AAAA records are locally available but they _aren't_ part of this zone.

Bottom line: we should do IP glue or IP hints, but not both.  Overall, my preference would be to stick with hints and get rid of the whole concept of glue.

File: draft-dnsop-deleg-transport.md [hunk 3]
```diff
@@ -182,8 +188,8 @@ To avoid a circular dependency, "tlsa" MUST appear in any DELEG record that is u
 Resolvers that support TLS-based transports MUST adopt one of the following behaviors:
 
 1. Use DANE for authentication, and treat any endpoints lacking DANE support as incompatible.
-1. Use PKI for authentication, and treat any DANE-only endpoint as incompatible.  In this behavior, compatibility with `tlsa=disabled` endpoints is REQUIRED.
-1. Support both DANE and PKI for authentication, preferring DANE if it is available for each endpoint.
+2. Use PKI for authentication, and treat any DANE-only endpoint as incompatible.  In this behavior, compatibility with `tlsa=disabled` endpoints is REQUIRED.
+3. Support both DANE and PKI for authentication, preferring DANE if it is available for each endpoint.
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:196]: Markdown auto-increments numeric lists, so the numbers here have no effect.

File: draft-dnsop-deleg-transport.md [hunk 4]
```diff
@@ -100,76 +100,82 @@ contributor:
...
+This document defines in detail the resolution process using these records.
 
-### SvcParams
+## SvcParams
 
-All SvcParamKeys for the "dns" scheme {{?9461}} apply as specified.  These are the "transport parameters", describing how to reach an endpoint.
+All SvcParamKeys for the "dns" scheme from SVCB-DNS {{?RFC9461}} apply as specified.  These are the "transport parameters", attributes describing how to reach an endpoint.
 
-The "alpn" transport parameter is OPTIONAL to include (unlike in SVCB-DNS, where it is generally required).  If the "alpn" SvcParamKey is omitted, the only available transport is presumed to be unencrypted DNS over UDP/TCP port 53.  Endpoints can indicate that insecure transport is not available by specifying "mandatory=alpn".
+The "alpn" transport parameter is OPTIONAL to include, unlike in SVCB-DNS, where it is generally required.  If the "alpn" SvcParamKey is omitted, the only available transport is presumed to be unencrypted DNS over UDP/TCP port 53.  Endpoints can indicate that insecure transport is not available by specifying "mandatory=alpn". [Why do we not require do53 to be explicit?]
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:176]: Are you arguing that Do53 should be opt-in instead of opt-out?  We can certainly structure it that way; the difference seems pretty superficial.  I appreciate the current approach as an application of Occam's Razor, but it's possible that "mandatory=alpn" is too obscure or is stretching the "mandatory" semantics too far.

File: draft-dnsop-deleg-transport.md [hunk 5]
```diff
@@ -196,6 +202,9 @@ alpn=doq tlsa="..." mandatory=alpn
 ;; services.  All resolvers can access it, but only resolvers with
 ;; DANE support will use DoQ.
 alpn=doq tlsa="..."
+[... how does this signal that it also offers do53? Is the though that it is
+  because only this draft specifies alpn and the core draft doesn't (currently)
+  have a method to explicitly signal do53?  Maybe it should... ]
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:211]: This server supports Do53 because it is not explicitly disclaimed by `mandatory=alpn`, as in the last example here.

File: draft-dnsop-deleg-transport.md (hunk 1 above)
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:226]: I don't understand this line.  Each ServiceMode record documents the supported transports for a particular endpoint, and the SvcPriority documents the recommended trial order.  Incompatible endpoints are skipped.  Of course clients are free to reorder them based on additional heuristics.  This is all covered in RFC 9460.

File: draft-dnsop-deleg-transport.md [hunk 6]
```diff
@@ -208,31 +217,33 @@ alpn=doq tlsa=disabled mandatory=alpn
...
-### Availability
+Thus, as with authority server selection for NS records, a resolver is free to use its own policy implementation to determine which servers to continue to use for resolution.
 
-If a zone operator removes all NS records before DELEG and SVCB records are implemented by all clients, the availability of their zones will be impacted for the clients that are using non-supporting resolvers. In some cases, this may be a desired quality, but should be noted by zone owners and operators.
+If a connection error occurs when a resolver attempts to access a nameserver delegated to by a DELEG or SVCB record, such as with a certificate mismatch or by being unreachable, the resolver SHOULD attempt to connect to the other nameservers in the DELEG/SVCB RRset to until either exhausting the list or the resolver's policy indicates that they should treat the resolution as failed.
 
-# Privacy Considerations
+For resolvers that strongly favor privacy, operators may wish to return a SERVFAIL when all attempts to use privacy-enhanced alpns in the DELEG/SVCB resolution process have failed. More permissive resolvers can fall back to using the legacy DNS resolution process via NS records.
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md]: I think this needs adjustment.  In this model, if I publish 5 DELEG records, and only one of them offers an encrypted endpoint, that single endpoint has to uphold my entire SLO.  (This could easily happen when delegating to multiple operators, or during a gradual rollout.)
```suggestion
Operators MAY return SERVFAIL when all attempts to use the DELEG/SVCB resolution process have failed.  This might have privacy benefits, especially if the DELEG records only offer encrypted transports.  More permissive resolvers can fall back to using the legacy DNS resolution process via NS records.
```

File: draft-dnsop-deleg-transport.md [hunk 7]
```diff
@@ -208,31 +217,33 @@ alpn=doq tlsa=disabled mandatory=alpn
...
 
-All of the information handled or transmitted by this protocol is public information published in the DNS.
+## Deployment Considerations
 
-# Security Considerations
+DELEG and SVCB records are intended to be the best method for modern resolvers to use delegation information for zones.
 
-TODO: Fill this section out
+Realistically, however, legacy resolvers will continue to exist on the Internet for a long time to come.  If a zone operator removes all NS records, the availability of their zones will be impacted for the clients that are using non-supporting resolvers. In some cases, this may be a desired quality, but should be noted by zone owners and operators.
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:238]: Why is this in the transport draft?  This topic seems like it should go in the main DELEG draft.

File: draft-dnsop-deleg-transport.md [hunk 8]
```diff
@@ -208,31 +217,33 @@ alpn=doq tlsa=disabled mandatory=alpn
...
 
-## Resolution procedure
+# Privacy Considerations
 
-TODO: Define how resolver selects which server connect to, especially when faced with selection of multiple transports. I's probably resolver's policy decision, but it should be said explicitly.
+All of the information transmitted by this protocol is public information published in the DNS.
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:242]: I don't think queries are meant to be "published" in the DNS.

File: draft-dnsop-deleg-transport.md [hunk 9]
```diff
@@ -208,31 +217,33 @@ alpn=doq tlsa=disabled mandatory=alpn
...
+
+# Security Considerations
 
-### Connection Failures
+No new security risks are introduced by this specification.  [Too bold to say?]
```
Comment (bemasc) [inline draft-dnsop-deleg-transport.md:246]: Sure, maybe we aren't making anything worse (although I worry about DNS Water Torture-like attacks), but security and privacy considerations still need to cover how the protocol can fail to achieve the intended improvements.  In this case, we need to talk about how TLS is authenticated, especially in unsigned zones.  (Related: #5)

**Review by bemasc — CHANGES_REQUESTED** [review]

File: draft-dnsop-deleg-transport.md (hunk 2 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:146]: > is only retrieved by an explicit query (QTYPE=DELEG)

That's not really a settled matter.  I still think, pending testing to confirm that for the vast majority of the resolver base, that the real deployment win here is to serve NS, DS and DELEG together for any delegation and NOT require an explicit query for them.  We're trying to make the whole business of finding modern auths more efficient, not add more queries. 

And I really, really don't think we should be adding addresses for the DELEG names to the Additional section, because something that understands what DELEG is at all should be using the hints.

File: draft-dnsop-deleg-transport.md (hunk 3 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:196]: I thought so but I was editing on a plane and had no formatter so couldn't confirm.  Just out of curiosity, if you have numbered them but done so (apparently) incorrectly, does it fix them?  eg, 1, 2, 2, 4 ...

File: draft-dnsop-deleg-transport.md (hunk 4 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:176]: Yes, I am arguing that if you are using a DELEG record, it should be explicit when you want do53 with it.  I don't want really anything assumed.  

It doesn't have to be in this draft initially though, since I can argue my case on the list.

(If all you wanted was Legacy DNS then just NS records.)

File: draft-dnsop-deleg-transport.md (hunk 5 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:211]: Okay, right, see above.  I think we should require signalling do53 to be explicit.

File: draft-dnsop-deleg-transport.md (hunk 1 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:226]: What don't you understand then?  It's just making it clear in this document as well.   Do you think this statement is inconsistent with that, or should just be shorter in drawing attention to it?

File: draft-dnsop-deleg-transport.md (hunk 8 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:242]: I also thought the sentence as a little awkward, but was working with some prior text.   I agree regarding queries.   This sentence could maybe be improved with "transmitted by authoritative servers", but the section needs work anyway.

File: draft-dnsop-deleg-transport.md (hunk 7 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:238]: Agreed.  It was here because it was the document I was working on first on the airplane, and I was addressing "fill this section out" with relevant thoughts.

File: draft-dnsop-deleg-transport.md (hunk 9 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:246]: Sounds good.  I wasn't really pushing for that sentence so much as putting a starting point that would draw out the "you were wrong on the Internet, and let me tell you why ..." brainstorming.

File: draft-dnsop-deleg-transport.md (hunk 1 above)
Comment (vttale) [reply draft-dnsop-deleg-transport.md:226]: I'm ambivalent on it.   While we still would need a way to distinguish AliasMode from ServiceMode, I'm kind of neutral about using a special magic 0 priority to do so.  I can see some limited use for it, but I also think it might be an attractive nuisance.  "Oh," says the Black hat,  "so the put this service at a lower priority?  Let's make that our  target!"

File: draft-dnsop-deleg-transport.md (hunk 3 above)
Comment (bemasc) [reply draft-dnsop-deleg-transport.md:196]: Yep, it completely ignores the actual numbers in the source code.  (This makes me sad: I want to be able to start with 0!)

File: draft-dnsop-deleg-transport.md (hunk 1 above)
Comment (bemasc) [reply draft-dnsop-deleg-transport.md:226]: OK.  If this is just meant to be highlighting some relevant aspects of RFC 9460, then it should say so.

Comment (vttale) [reply draft-dnsop-deleg-transport.md:226]: Will do.  I'll be working on some more editing tomorrow and will add the explicit reference.

**Comment by vttale** [comment]
Could we maybe get this PR merged and edit from there?

I'll break future edits into smaller PRs

File: draft-dnsop-deleg-transport.md (hunk 2 above)
Comment (bemasc) [reply draft-dnsop-deleg-transport.md:146]: It seems like this is not settled, and is also not relevant to the transport draft.  Please remove it or move it to the other draft.

BTW, using QTYPE=DELEG doesn't add queries.  It just replaces the "arbitrary" QTYPE in QNAME minimization with a meaningful one.  Non-DELEG servers _should_ treat it as an unknown QTYPE and return the NS (and glue) in a delegation response.

File: draft-dnsop-deleg-transport.md (hunk 4 above)
Comment (bemasc) [reply draft-dnsop-deleg-transport.md:176]:
```suggestion
The "alpn" transport parameter is OPTIONAL to include, unlike in SVCB-DNS, where it is generally required.  If the "alpn" SvcParamKey is omitted, the only available transport is presumed to be unencrypted DNS over UDP/TCP port 53.  Endpoints can indicate that insecure transport is not available by specifying "mandatory=alpn". [Should we change this to make do53 require a more explicit opt-in?]
```

File: draft-dnsop-deleg-transport.md (hunk 5 above)
Comment (bemasc) [reply draft-dnsop-deleg-transport.md:211]:
```suggestion
[ If we move to requiring an explicit opt-in for do53 then these examples will have to change. ]
```

File: draft-dnsop-deleg-transport.md (hunk 7 above)
Comment (bemasc) [reply draft-dnsop-deleg-transport.md:238]: OK, please delete this section.

File: draft-dnsop-deleg-transport.md (hunk 8 above)
Comment (bemasc) [reply draft-dnsop-deleg-transport.md:242]:
```suggestion
Negotiation of alternative transports with DELEG allows the use of authenticated encryption between resolvers and authoritative DNS servers.  This improves privacy by concealing the QNAME and QTYPE from a network observer.  This is especially beneficial when a resolver is only shared by a small number of users.
```

File: draft-dnsop-deleg.md [hunk 10]
```diff
@@ -104,19 +109,19 @@ contributor:
...
 
-The record lives in harmony with the existing NS records and any DS records used to secure the delegation with DNSSEC, with all provided in the Authority section of DNS responses.  Legacy DNS resolvers would continue to use the NS and DS records, while resolvers that understand DELEG and its associated parameters can efficiently switch.  [This is our proposed best-case scenario, but still needs testing to confirm the assertion about legacy resolvers.  We have several other backup plans about how it could be facilitated if the Authority method proves troublesome.]
+The record lives in harmony with the existing NS records and any DS records used to secure the delegation with DNSSEC, with all three types provided in the Authority section of DNS responses.  Legacy DNS resolvers will ignore the unknown DELEG type and continue to use the NS and DS records, per the testing described in Appendix A.  Resolvers that understand DELEG and its associated parameters can efficiently switch to the new mechanism.
```
Comment (bemasc) [inline draft-dnsop-deleg.md]:
```suggestion
The record lives in harmony with the existing NS records and any DS records used to secure the delegation with DNSSEC, with all three types provided in the Authority section of DNS responses.  Legacy DNS resolvers will ignore the unknown DELEG type and continue to use the NS and DS records, per the testing described in {{Testing}}.  Resolvers that understand DELEG and its associated parameters can efficiently switch to the new mechanism.
```

**Review by bemasc — CHANGES_REQUESTED** [review]

**Comment by paulehoffman** [comment]
A lot in the current draft has changed out from underneath this pull request. It might be better to close it and re-open it for just the bits that are still relevant.
